open_city_profile/middlewares.py

```diff
@@ -0,0 +1,33 @@
+from profiles.loaders import (
+    AddressesByProfileIdLoader,
+    EmailsByProfileIdLoader,
+    PhonesByProfileIdLoader,
+    PrimaryAddressForProfileLoader,
+    PrimaryEmailForProfileLoader,
+    PrimaryPhoneForProfileLoader,
+)
+
+LOADERS = {
+    "addresses_by_profile_id_loader": AddressesByProfileIdLoader,
+    "emails_by_profile_id_loader": EmailsByProfileIdLoader,
+    "phones_by_profile_id_loader": PhonesByProfileIdLoader,
+    "primary_address_for_profile_loader": PrimaryAddressForProfileLoader,
+    "primary_email_for_profile_loader": PrimaryEmailForProfileLoader,
+    "primary_phone_for_profile_loader": PrimaryPhoneForProfileLoader,
+}
+
+
+class GQLDataLoaders:
+    def __init__(self):
+        self.cached_loaders = {}
+
+    def resolve(self, next, root, info, **kwargs):
+        context = info.context
+
+        for loader_name, loader_class in LOADERS.items():
+            if loader_name not in self.cached_loaders:
+                loader = loader_class()
+                self.cached_loaders[loader_name] = loader
+                setattr(context, loader_name, loader)
```

Comment:
The flag could be an improvement 👍  I didn't see tha `cache_loaders` thing before. Seems like the loaders stay within the `context` on the same request, so need to cache them.

I'm not sure that having this as a middleware would be the craziest idea, since the same pattern is used to [authenticate the users with JWT](https://github.com/City-of-Helsinki/open-city-profile/blob/1cdcab40782e16fd0b0aeb20cbd6963afe7f10f9/open_city_profile/settings.py#L285), and I also don't know in which other non-hacky way we could initialise the context just once .

The [post](https://apirobot.me/posts/django-graphql-solving-n-1-problem-using-dataloaders) which has the clearest explanation adds a custom context to the `GraphQLView`. However, as you already said, the tests are the issue. Adding this context doesn't make it available to the tests.

Reply:
> I'm not sure that having this as a middleware would be the craziest idea, since the same pattern is used to [authenticate the users with JWT](https://github.com/City-of-Helsinki/open-city-profile/blob/1cdcab40782e16fd0b0aeb20cbd6963afe7f10f9/open_city_profile/settings.py#L285), and I also don't know in which other non-hacky way we could initialise the context just once .

This starts to get off topic, but I'll mention it anyway: a [recent commit](https://github.com/City-of-Helsinki/open-city-profile/commit/f727629308141a2c031ad6ba728487afbe793d4b) added another authentication system that doesn't do authentication in a resolver, partly because I thought it is a wrong way to do it. The old authentication system is still there and either one can be chosen to be used with a feature flag (not both of them simultaneously). In the future the older system can be removed.

But anyway, I'm ok with the middleware solution, for now at least.

Reply:
Ahhh I see now, I noticed the change but didn't pay much attention to how exactly the new authentication was done. The main issue, I think is the ability to re-use the functionality, both for code and testing. I'm strong on the opinion that you should not adapt your code to your tests, but the other way around. However, after few tries, the middleware approach was the most practical implementation.

If it's okay for you, we can go on with this for now. In any case, we'll keep looking for ways to improve as part of the Vene ⛵ backlog, so any improvements we can do, I'll try to also bring to Profile 👤 .

Reply:
Yes, we can go with the middleware way now.